Improve Zinc logging API

jvican opened the pull request:
This is the follow-up work to https://github.com/sbt/zinc/pull/304 and all the
comments that @stuhood has provided. Thank you for that.

Every commit is rich in details in its message. If you have any doubt about a
change, I invite you to read it.

I'm attaching a screenshot showing all the commits passing in Scala Center's
CI. I had to remove the first of the commits, and therefore all those passing
commits are lost.

jvican committed:
Use identity and remove unnecessary param

This does not change semantics.

jvican committed:
Fix #318: Add a `ReporterUtil` API

This API is inspired by the sbt backend that implements the whole
machinery for `ManagedLogger` and instantiates `LoggerReporter`.

The provided functionality is to get a default reporter configuration
that customises the default instantiation of the reporter. The
instantiation is provided in the same API.

We use contraband for the default reporter configuration to get the
builder pattern so that users can modify it more easily.

jvican committed:
Fix #319: Add a filtered compiler reporter

As explained in the issue and discussions [here](https://github.com/sbt/zinc/pull/304#issuecomment-309601296).

jvican committed:
Use `ReporterManager` for `AnalyzingCompiler`

jvican committed:
Add `getReporter` methods to `ReporterUtil`

jvican committed:
Add some code style changes

jvican committed:
Clean up DelegatingReporter

* Change `Reporter` to log problems (zxsbti.Problem`).
* Move `convert` to object (needs to be accessible from both logger and bridge).

jvican committed:
Remove use of `display` in `FIlteredReporter`

The use of display should be completely private, no matter who uses it.
It's weird that a logger has a concept of `display`. It's a bad
abstraction because the underlying logger could not "display".

Note that `display` in `LoggerReporter` has been made private despite
the fact that sbt uses it for printing warnings via `printWarnings`.

The use of this function reveals a small bug in the sbt codebase. In
`Defaults.scala`, sbt does the following:

``` def printWarningsTask: Initialize[Task[Unit]] = Def.task { val
analysis = compile.value match { case a: Analysis => a } val max =
maxErrors.value val spms = sourcePositionMappers.value val problems =
analysis.infos.allInfos.values.flatMap(i => i.getReportedProblems ++
i.getUnreportedProblems) val reporter = new LoggerReporter(max,
streams.value.log, foldMappers(spms)) problems foreach { p =>
reporter.display(p) } } ```

There's an error in the previous code since position mappers are not
used when "displaying" logs to the user, the use of the underlying
position mapper (provided by `foldMappers`) happens in `log`.

As we have already rewritten the `log` interface to accept `Problem`,
sbt can depend directly do instead:

``` problems foreach { p => reporter.display(p) } ```

which will solve that minor bug.

Note that there's a semantic difference between `display` and `log`.
`log` will persist every problem in a `problems` list. This however does
not change the semantics of the only caller of sbt since a new reporter
is created every time `printWarnings` is invoked, so that extra work
does not bother us.

jvican committed:
Remove sbt module dependencies in the bridge

The bridge has been compiled with the sbt module dependencies on the
classpath for ages. It seems that these dependencies trace back to the
time where Zinc was developed in sbt/sbt.

https://github.com/sbt/zinc/blame/4d4d83c636fb74516b6686f199fa19174adaea2d/build.sbt#L150

Having dependencies for the compiler bridge means two things:
* Users pay a penalty for resolving those artifacts.
* Compiler may be less efficient because of extra classpath entries (?)
* Zinc developers will be confused if they depend on something from
  those modules since bridge compilation will fail at runtime because
  dependencies will fail to link.

Previous to this change, the compiler bridge was being compiled with the
following classpath entries:

```
Vector(/tmp/sbt_af3b3cae/jars/org.scala-sbt/compiler-interface/compiler-interface-1.0.0-X16-SNAPSHOT.jar,
/tmp/sbt_af3b3cae/jars/org.scala-sbt/util-interface/util-interface-1.0.0-M24.jar,
/tmp/sbt_af3b3cae/jars/org.scala-sbt/io_2.12/io_2.12-1.0.0-M11.jar,
/tmp/sbt_af3b3cae/jars/org.scala-lang/scala-library/scala-library-2.12.1.jar,
/tmp/sbt_af3b3cae/jars/org.scala-sbt/util-logging_2.12/util-logging_2.12-1.0.0-M24.jar,
/tmp/sbt_af3b3cae/jars/org.scala-sbt/util-collection_2.12/util-collection_2.12-1.0.0-M24.jar,
/tmp/sbt_af3b3cae/jars/com.eed3si9n/sjson-new-core_2.12/sjson-new-core_2.12-0.7.1.jar,
/tmp/sbt_af3b3cae/jars/jline/jline/jline-2.14.3.jar,
/tmp/sbt_af3b3cae/jars/org.apache.logging.log4j/log4j-api/log4j-api-2.8.1.jar,
/tmp/sbt_af3b3cae/jars/org.apache.logging.log4j/log4j-core/log4j-core-2.8.1.jar,
/tmp/sbt_af3b3cae/jars/com.lmax/disruptor/disruptor-3.3.6.jar,
/tmp/sbt_af3b3cae/jars/com.eed3si9n/sjson-new-scalajson_2.12/sjson-new-scalajson_2.12-0.7.0.jar,
/tmp/sbt_af3b3cae/jars/org.mdedetrich/scala-json-ast_2.12/scala-json-ast_2.12-1.0.0-M6.jar,
/tmp/sbt_af3b3cae/jars/org.spire-math/jawn-parser_2.12/jawn-parser_2.12-0.10.4.jar,
/tmp/sbt_af3b3cae/jars/org.scala-lang/scala-reflect/scala-reflect-2.12.1.jar)
```

Right now, only two jars are brought in, and the two are Java
dependencies (the compiler interface and the util/logging interface).

jvican committed:
Add way to disable scripted buffer from sbt

jvican committed:
Remove need of `ManagedLogger` in `LoggerReporter`

https://github.com/sbt/zinc/pull/224/ introduced support for event
logging. However, this support forced downstream users depending on
`LoggerReporter` to provide a `ManagedLogger`. It also made it
impossible for Zinc users to create their own reporter because
`ManagedLogger` is part of the internal sbt API.

This commit factors out the event logging functionality out of the
`LoggerReporter`. Users that wish to pass a `ManagedReporter` must use
`ManagedLoggerReporter` which will set the logger up with all the
requirements to successfully log problems.

jvican committed:
Rename `LoggerReporter` to `LoggedReporter`

Since we have changed the signature of `LoggerReporter`, we don't want
our users to pass in a `ManagedLogger` that hasn't been properly set up
with the required event codecs.

Currently, people depending on `LoggerReporter` will assume that logger
reporter will set the logger up, but now it will not, since its
signature has been changed to accept a `xsbti.Logger`, which is a super
type of `ManagedLogger`. As the types match, typechecking will work and
people will find crappy exceptions at runtime because their
`ManagedLogger`s have not been set up.

Therefore, we change the name from `LoggerReporter` to `LoggedReporter`
to avoid this scenario. I also find the new name a little bit more
meaningful that the combination of two nouns. It's clear that it's a
reporter that uses an underlying log.

jvican committed:
Add a `ManagedFilteredReporter`

The managed filtered reporter is the counterpart of filtered reporter
but appropriately setting up a managed logger.

jvican committed:
Add public API to get reporter from Logger

Contrary to the previous API, now it's possible to get a reporter out of
a dumb `xsbti.Logger` that any user of the API can easily extend. This
new method covers almost all the needs that third-party developers need
to get Zinc working.

typesafe-tools commented:
To validate this pull request against other sbt modules, visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">THIS LINK.</A>

jvican commented:
![scalacenter-zinc-loggers](https://user-images.githubusercontent.com/2462974/27628533-a3399a0e-5bef-11e7-865d-bb12040346b9.png)

jvican commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/FilteredReporter.scala:
```diff
@@ -0,0 +1,86 @@
+      filters.exists(_.findFirstIn(str).isDefined)
+
+    severity != Severity.Error && (
+      (pos.sourceFile.isPresent && isFiltered(fileFilters, pos.sourceFile.get.getPath)) ||
```
Note that this uses `getPath` (and not `getAbsolutePath`) on purpose.

jvican commented:
CI is passing, I have requested three reviews. I need two to merge, as usual. I would appreciate feedback.
  
@eed3si9n @dwijnand I have tried to remove the Travis hook with no luck. It still happens to kick in for my PRs. If you happen to know why or know how to remove it, I would appreciate if you do so.

dwijnand commented:
Disabled Travis CI for this repo.

typesafe-tools commented:
To validate this pull request against other sbt modules, visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">THIS LINK.</A>

jvican commented:
It seems Travis CI is still enabled.

stuhood commented:
Looks good to me! Thanks @jvican .

eed3si9n reviewed (commented):
Overall it looks good to me.
Could you open a corresponding PR for sbt so either Dbuild or I can validate it.

jvican commented:
I'm happy to do it but I don't have access to dbuild. Can you give me permissions to do so?

eed3si9n commented:
Paging @cunei for Dbuild access.
For now, if you have a PR, I can run the validation.

cunei commented:
@eed3si9n any user can run validation now

typesafe-tools commented:
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/65/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: 1.0.x-pull/325/head-1.0-1.0-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/65/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | 1.0.x                | sbt/sbt@25d393bd8b34e3bf72d129c2427349c7aa881701
  zinc              | pull/325/head        | sbt/zinc@b6da4d506e6b5860d1598e1413548293b972d17d
  io                | 1.0                  | sbt/io@880de81bcc9bd4308885d2515d0a1b746026b946
  librarymanagement | 1.0                  | sbt/librarymanagement@7e87603b5fd3fa0d0ce38f6e85755d3973ddf114
  util              | 1.0                  | sbt/util@54cf4a43bb560d2e6fdc48fb51a376ca4228da03
  website           | 1.0.x                | sbt/website@5c32a0afcf64dc600ff154f5456026e2a74dc668

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=1.0.x&zinc=pull/325/head&io=1.0&librarymanagement=1.0&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

jvican commented:
@cunei Dbuild is not resolving dependencies.
  
@eed3si9n What's dbuild supposed to tell us? It's clear that this PR breaks sbt because the logger interface has changed. Is this something that you want to have for the posterity?

cunei commented:
At the moment the test is not really run by dbuild; some more work is needed to make it work properly with sbt 1.0.0-M5/6. The test is currently just an 'sbt compile' done by this script: https://github.com/sbt/sbt/blob/1.0.x/sbt-allsources.sh @dwijnand
You can specify in the build form multiple pull requests for the various components, and the script (and in the future dbuild) will check the changes in all the referenced modules together. For example: https://github.com/sbt/sbt/pull/3294#issuecomment-311832168

jvican commented:
@cunei Thanks for the explanation. My question really is what's the relation between dbuild and getting this PR merged. /cc @eed3si9n

eed3si9n commented:
@jvican See https://github.com/sbt/util/pull/85#issuecomment-311832175. This allows us to validate multiple PRs together and make sure that in conjunction they are able to compile sbt. This is to make sure that modules changes propagate within short window of time instead of catching breaking changes at the last minute.

jvican commented:
> This allows us to validate multiple PRs together and make sure that we in conjunction it's able to compile sbt.
  
Gotcha, so this would include sbt PRs too.
  
Can we please proceed with the merge of this PR? I have work on top of this and I need it to be in so that I can continue at a fast pace.

eed3si9n commented:
> Can we please proceed with the merge of this PR?

As I wrote 22h ago, could you open a corresponding PR for sbt so either Dbuild or I can validate it?

jvican commented:
> As I wrote 22h ago, could you open a corresponding PR for sbt so either Dbuild or I can validate it?
  
Didn't get what you were asking me. So you are asking me to create a PR to sbt migrating to my new Zinc changes. Okay.

eed3si9n commented:
Yea. FWIW I asked Martin to do the same for his watcher stuff too https://github.com/sbt/sbt/pull/3295

jvican commented:
@cunei Is dbuild also doing `Test/compile`?

cunei commented:
As I wrote, it's not really dbuild running in the validator, right now. At the moment the validator script it is only running 'sbt compile';my understanding is that 'sbt test', in that script, would not run the scripted tests. Once the real dbuild is working ok with 1.0.0-M6, validation will also run all the tests of the various modules.

jvican commented:
> As I wrote, it's not really dbuild running in the validator, right now. At the moment the validator script it is only running 'sbt compile';my understanding is that 'sbt test', in that script, would not run the scripted tests. Once the real dbuild is working ok with 1.0.0-M6, validation will also run all the tests of the various modules.
  
I'm asking if dbuild compiles the tests too (hence my previous reference to `Test/compile`). You said that dbuild is only doing `sbt compile` and you should compile tests too if you're compiling sources.

cunei commented:
No testing is currently done, just compilation of the main code. The tests are not currently being compiled, I believe.

jvican commented:
> No testing is currently done, just compilation of the main code. The tests are not currently being compiled, I believe.
  
Can you please compile tests too? Otherwise, we don't have the guarantee that the PR passes at least all the compilation steps.

cunei commented:
I changed `sbt compile` to `sbt ";compile;test:compile"`

dwijnand commented:
@cunei you can make that `sbt test` so that in addition to compiling everything you run the unit tests; scripted tests aren't run with `test`, they're run with `scripted`.

cunei commented:
@dwijnand OK, it is now running just `sbt-allsources.sh test`

Duhemm commented on internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala:
```diff
@@ -159,7 +159,7 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
   private[xsbt] def logUnreportedWarnings(seq: Seq[(String, List[(Position, String)])]): Unit = {
     val drep = reporter.asInstanceOf[DelegatingReporter]
```
I think that `drep` is not used anymore.

jvican replied:
Good catch, this can be removed, thanks :wink:

Duhemm commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/FilteredReporter.scala:
```diff
@@ -0,0 +1,86 @@
+ */
+class FilteredReporter(
+    fileFilters: Array[Regex],
+    msgFilters: Array[Regex],
```
How about `msgFilters: Array[String => Boolean]`?

Duhemm commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/ReporterManager.scala:
```diff
@@ -0,0 +1,102 @@
+      case Level.OFF     => sys.error("Level.OFF is not supported. Change the logging level.")
+      case _             => SbtLevel.Debug
+    }
+  }
```
Why don't we remove `sbt.util.Level` altogether then?

jvican replied:
I think it's too big of a change, doesn't give us much... historically we've always used sbt's level logging and I think it's good to honor that tradition. This is a workaround for a technical Java-Scala compatibility issue.

stuhood commented:
Very interested in seeing this one land (and making it into a release). I'm blocked for further testing of zinc 1.0.0-X18+ until it does.

jvican committed:
Address Martin's feedback

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">this link.</A>

jvican commented:
@Duhemm I've addressed all your comments.
  
@jastice You can have a look at this one now, it's been updated. The required changes concern only the compiler logger API in sbt/sbt, changes to get this in should be really small, sbt only needs `ManagedLoggedReporter`.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild/?zinc=pull/325/head&deploy_to_bintray=false">this link.</A>

typesafe-tools commented:
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/92/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/3316/head-pull/325/head-1.0-1.0-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/92/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3316">pull/3316/head      </A> | sbt/sbt@0cc3cfb36444e1efda7c4b5661f8a7ef95fcd7da
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/325">pull/325/head       </A> | sbt/zinc@510d64dd11f0593c592df257bb04cf7c3ee23449
  io                | <A HREF="https://github.com/sbt/io/tree/1.0">1.0                 </A> | sbt/io@f55849d988b94941309d316c43533b0af423f244
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.0">1.0                 </A> | sbt/librarymanagement@f8d3718d283cacdc11c6d4a9f101c675f687dc49
  util              | <A HREF="https://github.com/sbt/util/tree/1.0">1.0                 </A> | sbt/util@fe7743878fb590765582aec86d0e45df6ae0a936
  website           | <A HREF="https://github.com/sbt/website/tree/1.0.x">1.0.x               </A> | sbt/website@7e644bc8fc623589152e98bee75afd44893690e3

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=pull/3316/head&zinc=pull/325/head&io=1.0&librarymanagement=1.0&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

typesafe-tools commented:
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/93/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/3316/head-pull/325/head-1.0-pull/122/head-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/93/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3316">pull/3316/head      </A> | sbt/sbt@0cc3cfb36444e1efda7c4b5661f8a7ef95fcd7da
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/325">pull/325/head       </A> | sbt/zinc@510d64dd11f0593c592df257bb04cf7c3ee23449
  io                | <A HREF="https://github.com/sbt/io/tree/1.0">1.0                 </A> | sbt/io@f55849d988b94941309d316c43533b0af423f244
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/122">pull/122/head       </A> | sbt/librarymanagement@e3891c8732dc248bd3ff3be9d5c796454d66bc3a
  util              | <A HREF="https://github.com/sbt/util/tree/1.0">1.0                 </A> | sbt/util@fe7743878fb590765582aec86d0e45df6ae0a936
  website           | <A HREF="https://github.com/sbt/website/tree/1.0.x">1.0.x               </A> | sbt/website@7e644bc8fc623589152e98bee75afd44893690e3

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=pull/3316/head&zinc=pull/325/head&io=1.0&librarymanagement=pull/122/head&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>

typesafe-tools commented:
A dbuild validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/94/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/3316/head-pull/325/head-1.0-74d8a3835d535d3d787e9294539a0a3efc47674d-1.0-1.0.x)
dbuild <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/94/">has checked</A> the following projects against Scala 2.12:

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3316">pull/3316/head      </A> | sbt/sbt@0cc3cfb36444e1efda7c4b5661f8a7ef95fcd7da
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/325">pull/325/head       </A> | sbt/zinc@510d64dd11f0593c592df257bb04cf7c3ee23449
  io                | <A HREF="https://github.com/sbt/io/tree/1.0">1.0                 </A> | sbt/io@f55849d988b94941309d316c43533b0af423f244
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/74d8a3835d535d3d787e9294539a0a3efc47674d">74d8a3835d535d3d787e9294539a0a3efc47674d</A> | sbt/librarymanagement@74d8a3835d535d3d787e9294539a0a3efc47674d
  util              | <A HREF="https://github.com/sbt/util/tree/1.0">1.0                 </A> | sbt/util@fe7743878fb590765582aec86d0e45df6ae0a936
  website           | <A HREF="https://github.com/sbt/website/tree/1.0.x">1.0.x               </A> | sbt/website@7e644bc8fc623589152e98bee75afd44893690e3

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt%20validator/parambuild?sbt=pull/3316/head&zinc=pull/325/head&io=1.0&librarymanagement=74d8a3835d535d3d787e9294539a0a3efc47674d&util=1.0&website=1.0.x&deploy_to_bintray=false">(restart)</A>
